### ISSUE TITLE
The current get_clouds method will fail to fetch clouds in racing situation.

### ISSUE THREAD

**Issue by tjprescott** [Azure/azure-cli#3617]
Recent CI builds are failing randomly with the following error:
```
Failure: CloudNotRegisteredException (The cloud 'AzureCloud' is not registered.) 
```

Of the three Python version, with predictability, at least one of the versions is failing because of this.

Original title: Failure: CloudNotRegisteredException (The cloud 'AzureCloud' is not registered.) 

See my comments below including a repo

**Comment by tjprescott** [comment]
Just confirmed that this is also failing the master branch.

**Comment by mayurid** [comment]
@troydai : Thoughts?

**Comment by troydai** [comment]
I will go through this with @derekbekoe today. I investigated this issue last night. It occurs to me in parallel execution the `ConfigParser` will sometimes drop all the content of a config file.

**Comment by troydai** [comment]
When `get_clouds` is executed in high-parallelism context, there is a high chance it returns nothing. It is because the method will write to the config file first to ensure the known clouds were written and then read the config file again. There isn't lock to prevent and another process from reading the config file during writing eventually cause this issue.

Here's a simple repo demonstrate the root cause of this issue:
```Python
import os.path
import configparser
import multiprocessing
import sys

def init_clouds(config_file):
    config = configparser.ConfigParser()
    config.read(config_file)
    with open(config_file, 'w') as write_to:
        config.write(write_to)


def load_config(id):
    config_file = '/Users/troydai/.azure/clouds.config'
    init_clouds(config_file)

    if not os.path.exists(config_file):
        raise Exception('no file') 
    
    parser = configparser.ConfigParser()
    parser.read(config_file)
    sys.stdout.write(str(len(parser.sections())))

p = multiprocessing.Pool(2)
p.map(load_config, range(10000))
p.close()

p.join()
```

There is a larger implication here. It seems to me that we're frequently writing content to the config file. For each command execution, the same content is written to the configure file over and over and essentially erases many user settings. There are some modification users make will never take effect because of the force writing in every command.

**Comment by troydai** [comment]
I'll keep this issue open because the underlying issue is not addressed.

**Comment by troydai** [comment]
@derekbekoe I'd like to go through the design of fetching cloud again with you and Johan.

**Comment by mbiciunas** [comment]
My app is failing on this issue - trying to create 10 vm's in parallel in two regions (5 in each).  Based on above, adding a 5 sec. delay for each thread start seems to be working.  Of course, app now slower which is not so good.

BTW, running current azure-cli (2.0.7) on Python (Linux) 2.7.13 (default, Jan 12 2017, 17:59:37) 
[GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]

**Comment by mbiciunas** [comment]
Further testing shows same error occurring when creating security groups.  So I think I'll have to assume any concurrent activity is susceptible - which is not good.

If you come up with any workarounds / fixes, please post.

**Comment by troydai** [comment]
@mbiciunas one potential workaround is to leverage `AZURE_CONFIG_DIR` environment variable to customize the profile folder for the current process. When this env is not specified the az will look for profile credential in `~/.azure` folder. When you adopt this approach, you need to explicitly sign in for each process for automation.

**Comment by mbiciunas** [comment]
Thanks for your idea.  I ended up moving the cli calls to a separate method and implemented retry logic that checks the error message.  It's a workaround and seems to work.

This error will be an issue for anyone using cli in a script working with multiple resources at once.  So a fix is still really needed.

**Comment by listonb** [comment]
Slightly new edge case as well, High concurrency AZ command will corrupt the clouds.config file. Need to add a file lock on clouds.config.

**Comment by colemickens** [comment]
Just saw this in a Jenkins job. 38 good jobs, 1 bad one. No concurrency at play here though, only one job is active at a time and the `.azure` directory is local to a per-job-instance container anyway.

**Comment by troydai** [comment]
@colemickens if your case didn't happen under a concurrent situation, would you please open another issue and fill in more details? Thanks.
